# Lambda Functions

This type of functions allows you to write functions in a quick and potentially dirty way so I wouldn't advise you use them all the time but sometimes there are situations when they can come in a very handy . For example check the next example.

To declare a lambda function you have to specify the arguments of the function after of keyword **lambda**.

Compare lambda functions with convetional functions

In [58]:
# lambda way
sum1 = lambda x, y : x + y
sum1(3,4)


7

In [59]:
# conventional way
def sum2(x,y):
    return x+y
sum2(3,4)

7

In [60]:
%timeit sum1(100,50)

187 ns ± 36.6 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [61]:
%timeit sum2(100,50)

153 ns ± 21.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


**Exercise** : make the next conventional function one lambda function

In [62]:
def add_bangs1(word1):
    return word1 + "!!!"

add_bangs1('hello')

'hello!!!'

**Solution** :

In [63]:
add_bangs = (lambda a: a + '!!!')
add_bangs('hello')

'hello!!!'

### Anonymous functions:

## Map()

Function ´map()´ takes two arguments ´map( funct, seq). map() applies the function to All elements in the sequence. How the function, which will be applied to every value in the list (seq), doesn't has predefined name it will be one **anonymous function**.

The lambda function definition is: `add_bangs = (lambda a: a + '!!!')`, and the function call is: `add_bangs('hello')`.

In [64]:
x = [1,2,3,4,5,6]
squaredElements = map(lambda x : x**2, x)
print(list(squaredElements))

[1, 4, 9, 16, 25, 36]


In [65]:
def fahrenheit1(T):
    return ((float(9)/5)*T + 32)

def celsius1(T):
     return (float(5)/9)*(T-32)
 
temperatures = (36.5, 37, 37.5, 38, 39)
F = map(fahrenheit1, temperatures)
C = map(celsius1, F)

temperatures_in_Fahrenheit = list(map(fahrenheit1, temperatures))
temperatures_in_Celsius = list(map(celsius1, temperatures_in_Fahrenheit))
print(temperatures_in_Fahrenheit)

print(temperatures_in_Celsius)


[97.7, 98.60000000000001, 99.5, 100.4, 102.2]
[36.5, 37.00000000000001, 37.5, 38.00000000000001, 39.0]


In [66]:
F1 = list(map(lambda T: (float(9)/5)*T + 32, temperatures))
C2 = list(map(lambda T: (float(5)/9)*(T-32), F1))
print(F1)
print(C2)

[97.7, 98.60000000000001, 99.5, 100.4, 102.2]
[36.5, 37.00000000000001, 37.5, 38.00000000000001, 39.0]


map() can be applied to more than one list. The lists don't have to have the same length. map() will apply its lambda function to the elements of the argument lists, i.e. it first applies to the elements with the 0th index, then to the elements with the 1st index until the n-th index is reached: 

In [67]:
a = [1, 2, 3, 4]
b = [17, 12, 11, 10]
c = [-1, -4, 5]
list(map(lambda x, y, z : 2.5*x + 2*y - z, a, b, c))

[37.5, 33.0, 24.5]

Note: If one list has fewer elements than the others, map will stop when the shortest list has been consumed

The map function of the previous chapter was used to apply one function to one or more iterables. We will now write a function which applies a bunch of functions, which may be an iterable such as a list or a tuple, for example, to one Python object.

In [68]:
from math import sin, cos, tan, pi
#import math also works
import numpy as np
def map_functions(x, functions):
    """ map an iterable of functions on the the object x """
    res=[]
    for i in x:
        for func in functions:
            res.append(func(i))
    return res

In [69]:
values = np.linspace(0,pi,10)
family_of_functions = (sin, cos, tan)
print(map_functions(values, family_of_functions))

[0.0, 1.0, 0.0, 0.3420201433256687, 0.9396926207859084, 0.36397023426620234, 0.6427876096865393, 0.766044443118978, 0.8390996311772799, 0.8660254037844386, 0.5000000000000001, 1.7320508075688767, 0.984807753012208, 0.17364817766693041, 5.671281819617707, 0.984807753012208, -0.1736481776669303, -5.671281819617711, 0.8660254037844387, -0.4999999999999998, -1.7320508075688783, 0.6427876096865395, -0.7660444431189779, -0.8390996311772804, 0.3420201433256689, -0.9396926207859083, -0.36397023426620256, 1.2246467991473532e-16, -1.0, -1.2246467991473532e-16]


## Filtering
### Filter()

`filter(function, sequence)`

offers an elegant way to filter out all the elements of a sequence "sequence", for which the function function returns True. i.e. an item will be produced by the iterator result of filter(function, sequence) if item is included in the sequence "sequence" and if function(item) returns True. 

In other words: The function filter(f,l) needs a function f as its first argument. f has to return a Boolean value, i.e. either True or False. This function will be applied to every element of the list l. Only if f returns True will the element be produced by the iterator, which is the return value of filter(function, sequence). 


In [70]:
# Create a list of strings: fellowship
fellowship = ['frodo', 'samwise', 'merry', 'pippin', 'aragorn', 'boromir', 'legolas', 'gimli', 'gandalf']

# Use filter() to apply a lambda function over fellowship: result
result = filter(lambda a : len(a)>6, fellowship)

# Convert result to a list: result_list
result_list = list(result)


In the following example, we filter out first the odd and then the even elements of the sequence of the first 11 Fibonacci numbers: 

In [71]:
fibonacci = [0,1,1,2,3,5,8,13,21,34,55]
odd_numbers = list(filter(lambda x: x % 2, fibonacci))
print(odd_numbers)

even_numbers = list(filter(lambda x: x % 2 == 0, fibonacci))
print(even_numbers)

[1, 1, 3, 5, 13, 21, 55]
[0, 2, 8, 34]


## Reducing
### Reduce()

`reduce(func, seq)` 

continually applies the function func() to the sequence seq. It returns a single value. 

If seq = [ s1, s2, s3, ... , sn ], calling reduce(func, seq) works like this:

At first the first two elements of seq will be applied to func, i.e. func(s1,s2) The list on which reduce() works looks now like this: [ func(s1, s2), s3, ... , sn ]

In the next step func will be applied on the previous result and the third element of the list, i.e. func(func(s1, s2),s3)
The list looks like this now: [ func(func(s1, s2),s3), ... , sn ]

Continue like this until just one element is left and return this element as the result of reduce()


In [72]:
# Import reduce from functools
from functools import reduce

# Create a list of strings: stark
stark = ['robb', 'sansa', 'arya', 'brandon', 'rickon']

# Use reduce() to apply a lambda function over stark: result
result = reduce(lambda item1, item2 : item1 + item2, stark)

# Print the result

In [73]:
import functools
functools.reduce(lambda x,y: x+y, [47,11,42,13])

113

In [3]:
from functools import reduce
f = lambda a,b: a if (a > b) else b
reduce(f, [47,11,42,102,13])
102

102

Imagine an accounting routine used in a book shop. It works on a list with sublists, which look like this: 

| Order Number | Book Title and Author | Quantity | Price per Item 
| ---------- | ---------- | ---------- | ---------- 
| 34587 |	Learning Python, Mark Lutz |	4 |	40.95
| 98762 | Programming Python, Mark Lutz |	5 |	56.80 
| 77226 |	Head First Python, Paul Barry |	3 |	32.95 
| 88112 |	Einführung in Python3, Bernd Klein |	3 |	24.99 

Write a Python program, which returns a list with 2-tuples. Each tuple consists of a the order number and the product of the price per items and the quantity. The product should be increased by 10,- € if the value of the order is smaller than 100,00 €. 
Write a Python program using lambda and map.


The same bookshop, but this time we work on a different list. The sublists of our lists look like this: 
[ordernumber, (article number, quantity, price per unit), ... (article number, quantity, price per unit) ] 
Write a program which returns a list of two tuples with (order number, total amount of order).

In [75]:
orders = [ ["34587", "Learning Python, Mark Lutz", 4, 40.95], 
	       ["98762", "Programming Python, Mark Lutz", 5, 56.80], 
           ["77226", "Head First Python, Paul Barry", 3,32.95],
           ["88112", "Einführung in Python3, Bernd Klein", 	3, 24.99]]

min_order = 100
prices = list(map(lambda x: (x[0],x[2] * x[3]), orders))
print("prices without modfication: ", prices, "\n")
invoice_totals = list(map(lambda x: x if x[1] >= min_order else (x[0], x[1] + 10), prices))
print("prices with modfication: ",invoice_totals)

prices without modfication:  [('34587', 163.8), ('98762', 284.0), ('77226', 98.85000000000001), ('88112', 74.97)] 

prices with modfication:  [('34587', 163.8), ('98762', 284.0), ('77226', 108.85000000000001), ('88112', 84.97)]


The output of the previous program looks like this:
[('34587', 163.8), ('98762', 284.0), ('77226', 108.85000000000001), ('88112', 84.97)]

In [5]:
orders = [ [1, ("5464", 4, 9.99), ("8274",18,12.99), ("9744", 9, 44.95)], 
	       [2, ("5464", 9, 9.99), ("9744", 9, 44.95)],
	       [3, ("5464", 9, 9.99), ("88112", 11, 24.99)],
           [4, ("8732", 7, 11.99), ("7733",11,18.99), ("88112", 5, 39.95)] ]

min_order = 100
#print(orders[0][1:],"\n")
#print(list(map(lambda y: print(y[1]), orders)))
invoice_totals1 = list(map(lambda x: [x[0]] + list(map(lambda y: y[1]*y[2], x[1:])), orders))
print(" primer lista",invoice_totals1,"\n")
invoice_totals2 = list(map(lambda x: [x[0]] + [reduce(lambda a,b: a + b, x[1:])], invoice_totals1))
print("venta total",invoice_totals2)
invoice_totals3 = list(map(lambda x: x if x[1] >= min_order else (x[0], x[1] + 10), invoice_totals2))#less than 100 have 10 more
print("ventaja",invoice_totals3)
#print (invoice_totals)

 primer lista [[1, 39.96, 233.82, 404.55], [2, 89.91, 404.55], [3, 89.91, 274.89], [4, 83.93, 208.89, 199.75]] 

venta total [[1, 678.3299999999999], [2, 494.46000000000004], [3, 364.79999999999995], [4, 492.57]]
ventaja [[1, 678.3299999999999], [2, 494.46000000000004], [3, 364.79999999999995], [4, 492.57]]


We will get the following result:
[[1, 678.3299999999999], [2, 494.46000000000004], [3, 364.79999999999995], [4, 492.57]]

## Error Handling

We have statements introduced by an "except" keyword in Python. This keyword is responsible for capturing error raised from the operation we are performing. The main way to catch the exception is through try-except clause. Let's look how to capture any execption raised:

### try-except statement

In [77]:
def squared(x):
    """Return a square root of a number"""
    try:
        return x**(1/2)
    except:
        print("X must be an int or float")

In [78]:
squared("hi")

X must be an int or float


It's possible to create "custom-made" exceptions: With the raise statement it's possible to force a specified exception to occur.
Let's look at a simple example. Assuming we want to ask the user to enter an integer number. If we use a input(), the input will be a string, which we have to cast into an integer. If the input has not been a valid integer, we will generate (raise) a ValueError. We show this in the following interactive session:

In [79]:
n = int(input("Please enter a number: "))

Please enter a number: 56


With the aid of exception handling, we can write robust code for reading an integer from input:

In [80]:
def tryError():
    while True:
        try:
            n = input("Please enter an integer: ")
            n = int(n)
            break
        except ValueError:
            print("No valid integer! Please try again ...")
    print("Great, you successfully entered an integer!")

In [81]:
tryError()

Please enter an integer: 54
Great, you successfully entered an integer!


The while loop is entered. The code within the try clause will be executed statement by statement. If no exception occurs during the execution, the execution will reach the break statement and the while loop will be left. If an exception occurs, i.e. in the casting of n, the rest of the try block will be skipped and the except clause will be executed. The raised error, in our case a ValueError, has to match one of the names after except. In our example only one, i.e. "ValueError:". After having printed the text of the print statement, the execution does another loop. It starts with a new input(). 

### raise

More often than not, instead merely prining than error message, we will want to actually raise an error by using the keyword `raise`.  

In [82]:
squared(-9)# we won´t deal wit complex numbers

(1.8369701987210297e-16+3j)

In [6]:
def squared(x):
    """return squared root of a non-negative number"""
    if x<0:
        raise ValueError("X must be non-negative number")
    try:
        return x**(1/2)
    except TypeError:
        print("X must be an integer or float")

In [84]:
squared(-9)

ValueError: X must be non-negative number

### Multiple Except Clauses

A try statement may have more than one except clause for different exceptions. But at most one except clause will be executed.

Our next example shows a try clause, in which we open a file for reading, read a line from this file and convert this line into an integer. There are at least two possible exceptions:

* an IOError
* ValueError

Just in case we have an additional unnamed except clause for an unexpected error:

In [7]:
def readNumber(path):
    number = []
    try:
        with open(path, 'r') as f:
            for line in f:
                for i in line.split(','):# remove the character given by parameter
                    number.append(int(i))
        return number
    except IOError as e:
        #errornumber, message = e.args
        #print("I/O error({0}): {1}".format(errornumber,message))
        # e can be printed directly without using .args:
        print(e)
    except ValueError:
        print("No valid integer in line.")

In [149]:
path = "./numbers.txt"
readNumber(path)

[11, 2, 3, 4, 5, 6, 65, 7654, 8483, 2536, 57]

In [150]:
path = "doesn't exist"
readNumber(path)

[Errno 2] No such file or directory: "doesn't exist"


In [152]:
path = 123
readNumber(path)

[WinError 6] Controlador no válido


In [142]:
path = "./chistes.txt"
readNumber(path)

No valid integer in line.


### Clean-up Actions (try ... finally)
The try statement can be followed by a finally clause. Finally clauses are called clean-up or termination clauses, because they must be executed under all circumstances, i.e. a "finally" clause is always executed regardless if an exception occurred in a try block or not. 

In [154]:
def readNumber(path):
    number = []
    try:
        with open(path, 'r') as f:
            for line in f:
                for i in line.split(','):# remove the character given by parameter
                    number.append(int(i))
        return number
    except IOError as e:
        #errornumber, message = e.args
        #print("I/O error({0}): {1}".format(errornumber,message))
        # e can be printed directly without using .args:
        print(e)
    except ValueError:
        print("No valid integer in line.")
    finally:
        number.append(None)
        return number

In [155]:
path = "./chistes.txt"
readNumber(path)

No valid integer in line.


[None]

In [ ]:
def readNUmber(path):
    number = []
    total = []
    try:
        with open(path,'r') as f:
            for line in f:
                for i in line.split(','):# remove the character given by parameter
                    number.append(float(i))
                    if len(number)>=4:
                        total.append((number[0],((number[1]*number[2])*(1 - number[3])))) 
                        number= []   
        return total
    except IOError as e:
        print(e)
    except ValueError:
        print("no valid integer in line")
    finally:
        number.append([1,2,3])
        return total

In [23]:
path = "./numbers2.txt"
readNumber(path)

[(1.0, 1080.0), (1.0, 1080.0)]